File: be/user_be_java/src/main/java/com/user/service/UserService.java
package com.user.service;

import com.user.domain.User;
import com.user.dto.PageResponse;
import com.user.dto.UserCreateDTO;
import com.user.dto.UserQueryDTO;
import com.user.dto.UserUpdateDTO;
import com.user.common.exception.AppException;
import com.user.mapper.UserMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.BeanUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

/**
 * 用户服务类
 */
@Service
@RequiredArgsConstructor
public class UserService {

    @Autowired
    private final UserMapper userMapper;

    @Value("${app.be-type}")
    private String beType;

    /**
     * 分页查询用户列表
     */
    public PageResponse<User> getUserWithPageAndCount(Integer page, Integer limit, UserQueryDTO queryDTO) {
        // 计算偏移量
        int offset = (page - 1) * limit;
        
        // 查询数据
        List<User> users = userMapper.selectUsersWithConditions(
                queryDTO.getName(),
                queryDTO.getEmail(),
                queryDTO.getAge(),
                queryDTO.getBeType(),
                queryDTO.getStartDate(),
                queryDTO.getEndDate(),
                offset,
                limit
        );
        
        // 统计总数
        long total = userMapper.countUsersWithConditions(
                queryDTO.getName(),
                queryDTO.getEmail(),
                queryDTO.getAge(),
                queryDTO.getBeType(),
                queryDTO.getStartDate(),
                queryDTO.getEndDate()
        );
        
        // 计算总页数
        int totalPages = (int) Math.ceil((double) total / limit);
        
        return new PageResponse<>(users, total, page, limit, totalPages);
    }

    /**
     * 根据 ID 获取用户
     */
    public User getUserById(Integer id) {
        User user = userMapper.selectById(id);
        if (user == null) {
            throw new AppException(404, "User not found");
        }
        return user;
    }

    /**
     * 创建用户
     */
    @Transactional
    public User createUser(UserCreateDTO dto) {
        // 检查邮箱是否已存在
        if (userMapper.existsByEmail(dto.getEmail())) {
            throw new AppException(409, "Email already exists");
        }

        User user = new User();
        BeanUtils.copyProperties(dto, user);
        user.setAge(dto.getAge() != null ? dto.getAge() : 0);
        user.setBeType(beType); // 设置后端类型
        user.setCreatedAt(LocalDateTime.now());
        user.setUpdatedAt(LocalDateTime.now());

        userMapper.insert(user);
        return user;
    }

    /**
     * 更新用户
     */
    @Transactional
    public User updateUser(Integer id, UserUpdateDTO dto) {
        User user = userMapper.selectById(id);
        if (user == null) {
            throw new AppException(404, "User not found");
        }

        // 如果邮箱改变，检查新邮箱是否已存在
        if (dto.getEmail() != null && !dto.getEmail().equals(user.getEmail())) {
            if (userMapper.existsByEmail(dto.getEmail())) {
                throw new AppException(409, "email already exists");
            }
        }

        // 更新用户信息
        user.setName(dto.getName());
        user.setEmail(dto.getEmail());
        user.setAge(dto.getAge());
        user.setBirthDate(dto.getBirthDate());
        user.setUpdatedAt(LocalDateTime.now());

        userMapper.updateById(user);
        return user;
    }

    /**
     * 删除用户
     */
    @Transactional
    public void deleteUser(Integer id) {
        User user = userMapper.selectById(id);
        if (user == null) {
            throw new AppException(404, "User not found");
        }
        userMapper.deleteById(id);
    }
}

File: be/user_be_go/internal/service/user_service.go
package service

import (
	"math"
	"time"
	"user_be_go/internal/config"
	"user_be_go/internal/constants"
	"user_be_go/internal/domain"
	"user_be_go/internal/dto"
	"user_be_go/internal/repository"
	"user_be_go/internal/utils"
)

type UserService struct {
	repo *repository.UserRepository
}

func NewUserService(repo *repository.UserRepository) *UserService {
	return &UserService{repo: repo}
}

// GetUsers 获取所有用户
func (s *UserService) GetUsers() ([]domain.User, error) {
	return s.repo.FindAll()
}

// GetUserByID 根据ID获取用户
func (s *UserService) GetUserByID(id int) (*domain.User, error) {
	user, err := s.repo.FindByID(id)
	if err != nil {
		return nil, err
	}
	if user == nil {
		return nil, utils.NewNotFoundError(constants.MsgUserNotFound)
	}
	return user, nil
}

// GetUsersWithPageAndCount 分页查询用户
func (s *UserService) GetUsersWithPageAndCount(query dto.UserQueryDTO) (*dto.PageResponse, error) {
	users, total, err := s.repo.FindWithPageAndCount(query)
	if err != nil {
		return nil, err
	}

	totalPages := int(math.Ceil(float64(total) / float64(query.Limit)))

	return &dto.PageResponse{
		Data:       users,
		Total:      total,
		Page:       query.Page,
		Limit:      query.Limit,
		TotalPages: totalPages,
	}, nil
}

// CreateUser 创建用户
func (s *UserService) CreateUser(dto dto.UserCreateDTO) (*domain.User, error) {
	// 检查邮箱是否已存在
	existing, err := s.repo.FindByEmail(dto.Email)
	if err != nil {
		return nil, err
	}
	if existing != nil {
		return nil, utils.NewConflictError(constants.MsgUserEmailExists)
	}

	// 创建用户
	user := &domain.User{
		Name:      dto.Name,
		Email:     dto.Email,
		BirthDate: dto.BirthDate.ToTimePtr(),
		BeType:    config.AppConfig.BeType,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	if dto.Age != nil {
		user.Age = *dto.Age
	}

	if err := s.repo.Create(user); err != nil {
		return nil, err
	}

	return user, nil
}

// UpdateUser 更新用户
func (s *UserService) UpdateUser(id int, dto dto.UserUpdateDTO) (*domain.User, error) {
	// 检查用户是否存在
	user, err := s.repo.FindByID(id)
	if err != nil {
		return nil, err
	}
	if user == nil {
		return nil, utils.NewNotFoundError(constants.MsgUserNotFound)
	}

	// 如果更新邮箱，检查新邮箱是否已存在
	if dto.Email != nil && *dto.Email != user.Email {
		existing, err := s.repo.FindByEmail(*dto.Email)
		if err != nil {
			return nil, err
		}
		if existing != nil {
			return nil, utils.NewConflictError(constants.MsgUserEmailExists)
		}
		user.Email = *dto.Email
	}

	// 更新字段
	if dto.Name != nil {
		user.Name = *dto.Name
	}
	if dto.Age != nil {
		user.Age = *dto.Age
	}
	if dto.BirthDate != nil {
		user.BirthDate = dto.BirthDate.ToTimePtr()
	}
	user.UpdatedAt = time.Now()

	if err := s.repo.Update(id, user); err != nil {
		return nil, err
	}

	return user, nil
}

// DeleteUser 删除用户
func (s *UserService) DeleteUser(id int) error {
	// 检查用户是否存在
	user, err := s.repo.FindByID(id)
	if err != nil {
		return err
	}
	if user == nil {
		return utils.NewNotFoundError(constants.MsgUserNotFound)
	}

	return s.repo.Delete(id)
}
